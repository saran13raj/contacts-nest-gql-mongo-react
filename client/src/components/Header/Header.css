.header {
	position: sticky;
	top: 0;
	width: 100vw;
	/* background: linear-gradient(-270deg, rgb(87, 31, 188), rgb(221, 67, 44)); */
	background: linear-gradient(-270deg, #2d46b9, #f037a5);
	color: white;
	display: flex;
	justify-content: space-around;
	align-items: center;
	padding: 20px;
	height: 60px;
}

.header__home {
}

.header__title {
	text-decoration: none;
}

.header__title__text {
	color: #ffffff;
	margin: 0;
}

.header__action {
	/* margin-top: 2%; */
}
