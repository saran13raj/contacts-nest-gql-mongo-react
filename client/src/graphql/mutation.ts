import { gql } from '@apollo/client';

// mutation InsertContact($input: ContactInput!) {
// createContact(input: $input) {
export const INSERT_ONE_CONTACT = gql`
	mutation InsertContact($name: String!, $phone: String!, $email: String!, $picture: String!) {
		createContact(input: { name: $name, phone: $phone, email: $email, picture: $picture }) {
			_id
			name
			phone
			email
			picture
		}
	}
`;

export const DELETE_ONE_CONTACT = gql`
	mutation DeleteOneContact($_id: String!) {
		deleteContact(_id: $_id)
	}
`;

export const UPDATE_ONE_CONTACT = gql`
	mutation UpdateOneContact(
		$_id: String!
		$name: String!
		$phone: String!
		$email: String!
		$picture: String!
	) {
		updateContact(
			data: {
				query: { _id: $_id }
				set: { name: $name, phone: $phone, email: $email, picture: $picture }
			}
		) {
			_id
			name
			phone
			email
			picture
		}
	}
`;
