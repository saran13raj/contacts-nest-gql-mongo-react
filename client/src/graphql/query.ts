import { gql } from '@apollo/client';

export const FETCH_ALL_CONTACTS_QUERY = gql`
	query FetchAllContacts {
		contacts {
			_id
			name
			phone
			email
			picture
			info
		}
	}
`;

export const FECTH_ONE_CONTACT_QUERY = (_id: string) => gql`
query FetchOneContact($_id: String!){
    contact(_id: "${_id}") {
      _id
      name
      phone
      email
      picture
      info
    }
  }
`;
