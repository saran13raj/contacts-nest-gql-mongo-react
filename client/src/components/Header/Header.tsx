import React, { useState } from 'react';
import Modal from 'react-modal';
import { useMutation } from '@apollo/client';
import { Redirect } from 'react-router';
import { Link } from 'react-router-dom';

import './Header.css';
import { INSERT_ONE_CONTACT } from '../../graphql/mutation';

const modalStyles = {
	content: {
		height: '350px',
		width: '600px',
		marginTop: '100px',
		marginLeft: 'auto',
		marginRight: 'auto'
	}
};

Modal.setAppElement('#root');

const Header = () => {
	const [name, setName] = useState('');
	const [number, setNumber] = useState('');
	const [email, setEmail] = useState('');
	const [picture, setPicture] = useState('');
	const [showAddContact, setShowAddContact] = useState(false);
	const [showError, setShowError] = useState(false);

	const [addContact, { data, loading, error }] = useMutation(INSERT_ONE_CONTACT, {
		variables: {
			// input: {
			name,
			phone: number,
			email,
			picture
			// }
		}
	});

	if (loading) {
		console.log('Loading:::');
	}

	if (error) {
		console.log('Error adding contact:::', error);
	}

	const onAddContact = async (e: any) => {
		e.preventDefault();
		if (name.length === 0 || number.length === 0) {
			setShowError(true);
			return;
		} else {
			const contactObj = {
				name,
				number,
				email,
				picture
			};
			console.log(contactObj);

			await addContact();
			if (data) {
				<Redirect to='/' />;
			}

			setShowError(false);
			setShowAddContact(false);
		}
	};

	const errorMessage = () => {
		return <p className='header__errorMessage'>Please enter valid name and number</p>;
	};

	return (
		<div className='header'>
			<Link to='/' className='header__home'>
				<h3>Contacts app</h3>
			</Link>
			<div className='header__action'>
				<button onClick={() => setShowAddContact(true)}>Add contact</button>
			</div>
			<Modal isOpen={showAddContact} style={modalStyles}>
				<div className='header__add'>
					<div className='header__add__title'>
						<h2>Please enter the contact details</h2>
						<button onClick={() => setShowAddContact(false)}>Close</button>
					</div>
					<div className='header__add__form'>
						<form>
							<div className='header__add__detail'>
								<label>Name</label>
								<input
									type='text'
									value={name}
									onChange={(e) => setName(e.target.value)}
								/>
							</div>
							<div className='header__add__detail'>
								<label>Mobile Number</label>
								<input
									type='text'
									value={number}
									onChange={(e) => setNumber(e.target.value)}
								/>
							</div>
							<div className='header__add__detail'>
								<label>Email</label>
								<input
									type='text'
									value={email}
									onChange={(e) => setEmail(e.target.value)}
								/>
							</div>
							<div className='header__add__detail'>
								<label>Profile Picture</label>
								<input
									type='text'
									value={picture}
									onChange={(e) => setPicture(e.target.value)}
								/>
							</div>
							<div className='header__add__button__div'>
								<button onClick={onAddContact}>Add</button>
							</div>
							<div className='header__add__error'>
								{showError ? errorMessage() : null}
							</div>
						</form>
					</div>
				</div>
			</Modal>
		</div>
	);
};

export default Header;
