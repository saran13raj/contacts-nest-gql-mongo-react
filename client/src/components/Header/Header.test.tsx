import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { MemoryRouter } from 'react-router-dom';

import Header from './Header';
import { INSERT_ONE_ADDRESS, INSERT_ONE_CONTACT } from '../../graphql/mutation';

const mocks = [
	{
		request: {
			query: INSERT_ONE_ADDRESS,
			variables: {
				location: 'Chennai'
			}
		},
		result: {
			data: {
				createAddress: { id: '1', location: 'Chennai' }
			}
		}
	},
	{
		request: {
			query: INSERT_ONE_CONTACT,
			variables: {
				name: 'test',
				phone: '1234567890',
				email: 'test2g.com',
				picture: '',
				info: { age: 24, area: 'test area' },
				address: '1'
			}
		},
		result: {
			data: {
				name: 'test',
				phone: '1234567890',
				email: 'test2g.com',
				picture: '',
				info: { age: 24, area: 'test area' },
				address: {
					_id: '1',
					location: 'Chennai'
				}
			}
		}
	}
];

describe('Header component', () => {
	it('renders Header without crashing', () => {
		const component = TestRenderer.create(
			<MockedProvider mocks={mocks} addTypename={false}>
				<MemoryRouter>
					<Header />
				</MemoryRouter>
			</MockedProvider>
		);

		// console.log(component);
		const tree = component.toJSON();
		// console.log(tree);
		expect(tree).toMatchSnapshot();
		// console.log(tree?.props.children);
		// expect(tree.children).toContain('Contacts app')
	});
});
