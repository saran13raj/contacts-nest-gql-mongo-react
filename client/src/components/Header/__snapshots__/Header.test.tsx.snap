// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders Header without crashing 1`] = `
<div
  className="header"
>
  <div
    className="header__home"
  >
    <a
      className="header__title"
      href="/"
      onClick={[Function]}
    >
      <h2
        className="header__title__text"
      >
        Contacts app
      </h2>
    </a>
  </div>
  <div
    className="header__action"
  >
    <button
      className="ant-btn ant-btn-round"
      color="#D5D5D5"
      onClick={[Function]}
      type="button"
    >
      <span>
        Add contact
      </span>
    </button>
  </div>
</div>
`;
